ACRN SMP fixes

andyross opened the pull request:
Fixes needed to get Zephyr running with more than one CPU under the ACRN hypervisor.  Nothing really complicated, though the first patch is an ugly hack.

See earlier conversation in #35310 which contains a different version of the first patch.

Fixes #35076

Commit:
arch/x86_64: Terrible, awful hackery to bootstrap entry

Because of a historical misunderstanding, by default the ACRN
hypervisor wants to load Zephyr at address 0x1000 and enter the binary
at that same address.  This entry point corresponds to the __start
symbol of the build they were given, which is a 1-cpu non-SMP
configuration.  Unfortunately, when we build with
CONFIG_MP_NUM_CPUS=1, the code in locore.S #if's out the 16 bit entry
point for the auxiliary CPUs at the start of the section.  So in the
build ACRN received, the start address happened to be 0x7000, the same
address we need to launch the AP processors from.

That's right: under ACRN, the SAME ADDRESS used to enter the OS in 32
bit mode needs to be used later to boot CPUs running in 16 bit real
mode!

The solution, such as it is, is to put a 32 bit jump at the entry
address which hops to the 32 bit OS entry code, and then scribble NOP
instructions over that jump once we get there so that the next time we
reach that address (in real mode) we fall through to the correct
entry.

This patch should be considered a temporary workaround.  While it
works on all x86 hardware, it's not really needed.  A much better
solution would be to eliminate the locore linker region entirely
(which causes other headaches) and enter the Zephyr binary in a 32 bit
address somewhere in the contiguous high memory area.  All that locore
is needed for is the 16 bit bootstrap code for SMP processors, which
is ~6 instructions and can be copied in from the kernel at runtime.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86_64: Use modern CR0 assembly

The 16 bit bootstrap code for SMP CPUs was using the 286-era "lmsw"
instruction (load machine status word) to set the protected bit in CR0
(which is the modern evolution of the same register), presumably
because this is 16 bit code and we can't move a dword into CR0.

But that's wrong, because the full instruction set *is* available in
real mode on a 386, you just have to use a operand size prefix to get
to it, which the assembler emits for you automatically when you use
the .code16 directive.

Write this conventionally and use modern (e.g. 1986-era) instructions.
It also has the advantage of not confusing much more modern
hypervisors like ACRN by issuing instructions they (and I!) never knew
existed.

Fixes #35076

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/x86/acrn: Add APIC IDs

The ACRN hypervisor uses 0, 2, 4, 6 as its local APIC IDs for
virtualized CPUs and not the 0, 1, 2, 3 defaults we have.

(I hate this feature, having to manually (!) probe and code these
things in C isn't scaling.  Zephyr needs to do the probing on its own
somehow, even if it's an offline tool in Linux or something.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/x86/acrn_ehl_crb: Enable SMP by default

SMP is working now, make a 2-cpu configuration default for this
device.

Note that this requires changes to the default ACRN build
configuration!  In hybrid.xml, you need to specify multiple physical
CPUs for the VM to uses, e.g.:

    <vm id="0">
        ...
        <cpu_affinity>
            <pcpu_id>0</pcpu_id>
            <pcpu_id>1</pcpu_id>
        </cpu_affinity>
    </vm>

Failing to build with this change will result in the system hanging at
boot trying to start up a CPU that won't run.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/x86/acrn: Rework board documentation

ACRN build and configuration is non-trivially complicated, and so far
integration documentation has been mostly missing, and users have had
to get by via copying from existing integration efforts with minor
changes, leading to repeated mistakes and persistent confusion.  This
is an attempt to document the process from first principles, with an
eye toward informing integrators (not me!) who might come by later to
better automate things.  Some of the content is going to look remedial
to someone already familiar with e.g. ACRN configuration or EFI boot.

This simply replaces the pre-existing docs, which were for earlier
versions of ACRN where Zephyr was launched from the service OS instead
of the now-standard pre-launch VM mode.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
This is enough to get SMP working well for me.  I still see a few test failures here and there (notably the arch_nop test seems to be complaining that it takes zero cycles, which might be an emulation artifact), but basically everything runs as expected.  See the docs added for details, where I've tried to spell out explicitly all the pitfalls and footguns in the process.

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -124,8 +124,10 @@ x86_ap_start:
 	lgdt gdt48
 	lidt idt48
-	movw $1, %ax
-	lmsw %ax
```
Full disclosure, this may have been an instruction introduced in 1982, but until 2021-05-27 I did not know it existed.

stephanosio commented on arch/x86/core/intel64/locore.S:
```diff
@@ -78,15 +78,41 @@
 .section .locore,"ax"
+.code32
+.globl __start
+__start:
+	jmp __start32
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
```
Can't you just check if the proc is in protected mode using `smsw` and jump to `__start32`, using instruction encodings that are common to both real and protected modes? (after all, it doesn't matter whether we use `eax` or `ax`, since `PE` is the LSB).

stephanosio replied:
> The MSW register has other problems
or `mov eax, cr0` for that matter; it's `0F20C0` in both real and protected modes.

andyross replied:
Touching EAX in real mode requires an address size prefix, no?  I really don't see that kind of trickery as an improvement.  The goal of the code here is to do the right thing and be as small a footgun as possible until we get this fixed architecturally.  This code is ugly, but it works by expecting the assembler to generate normal code in the mode it's working in and the only instruction encoding fu required by the reader is understanding that 0x90 is a 1-byte NOP.

stephanosio replied:
Control register `mov` has a different encoding and doesn't require operand override prefix (66h).

I thought doing that would make things a lot simpler (and this kind of hack isn't all that uncommon either), but it's really up to you :)

andyross replied:
Sorry, but "This section of assembly gets run in both 32 and 16 bit mode.  Please only use instructions that encode the same in both modes. Kthx." just isn't going to seem "simpler" to anyone who hasn't memorized volume 2 of the IA SDM.  And again, longer term we want the whole locore section to go away.

Commit:
boards/x86/acrn: Use APIC_TSC_DEADLINE_TIMER

This board got forgotten when we migrated the older APIC_TIMER users.
Now the platform is SMP by default and the older driver refuses to
build.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

jenmwms commented:
Linking this to the issue: https://github.com/zephyrproject-rtos/zephyr/issues/35076

jhedberg commented:
> Linking this to the issue: #35076

Do you mean that merging this should close that issue? Please then add the appropriate "Fixes #..." to the PR description (AFAIK that's the only way that GitHub will understand the relation)

jenmwms commented:
> > Linking this to the issue: #35076
> 
> Do you mean that merging this should close that issue? Please then add the appropriate "Fixes #..." to the PR description 

Yes that's it. I thought @andyross needed to be the one to update the PR/commit message, but looks like I can do that too. Thanks :)

nashif commented on boards/x86/acrn/README.rst:
```diff
@@ -0,0 +1,230 @@
+Building and Running Zephyr with ACRN
```
This needs to go into `boards/x86/acrn/doc/` and either expand on the existing doc or replace it.

nashif commented on boards/x86/acrn/README.rst:
```diff
@@ -0,0 +1,230 @@
+
+        west build -b acrn_ehl_crb samples/hello_world
+
+Note the kconfig output in `build/zephyr/.config`, you will need to
```
add another backquote, so ``, this needs to be done for all single quotes..

andyross commented:
Rebase.  Fix up doc format (hopefully) and move the file.
